pkg/didcomm/protocol/introduce/service_test.go

```diff
@@ -568,8 +547,146 @@ func TestService_save(t *testing.T) {
+		checkStateMsg(t, sCh, service.PostState, ProposalMsgType, stateNameArranging)
+
+		respMsg1, err := service.NewDIDCommMsg(toBytes(t, Response{
+			Type:   ResponseMsgType,
+			ID:     uuid.New().String(),
+			Thread: &decorator.Thread{ID: thID},
+		}))
+		require.NoError(t, err)
+
+		// handle Response msg (receives Response)
+		go func() {
+			// nolint: govet
+			_, err := svc.HandleInbound(respMsg1)
+			require.NoError(t, err)
+		}()
+
+		dep := mocks.NewMockActionDependency(ctrl)
+		dep.EXPECT().Destinations().Return([]*service.Destination{
+			{ServiceEndpoint: "service/endpoint1"},
+			{ServiceEndpoint: "service/endpoint2"},
+		}).Times(2)
+
+		continueAction(t, aCh, ResponseMsgType, dep)
+		checkStateMsg(t, sCh, service.PreState, ResponseMsgType, stateNameArranging)
+		checkStateMsg(t, sCh, service.PostState, ResponseMsgType, stateNameArranging)
+
+		respMsg2, err := service.NewDIDCommMsg(toBytes(t, Response{
+			Type:   ResponseMsgType,
+			ID:     uuid.New().String(),
+			Thread: &decorator.Thread{ID: thID},
+		}))
+		require.NoError(t, err)
+
+		// handle Response msg (receives Response)
+		go func() {
+			// nolint: govet
+			_, err := svc.HandleInbound(respMsg2)
+			require.NoError(t, err)
+		}()
+
+		continueAction(t, aCh, ResponseMsgType, dep)
+		checkStateMsg(t, sCh, service.PreState, ResponseMsgType, stateNameDelivering)
+		checkStateMsg(t, sCh, service.PostState, ResponseMsgType, stateNameDelivering)
+		checkStateMsg(t, sCh, service.PreState, ResponseMsgType, stateNameDone)
+		checkStateMsg(t, sCh, service.PostState, ResponseMsgType, stateNameDone)
+	})
+
+	t.Run("Introducee", func(t *testing.T) {
+		svc, err := New(&protocol.MockProvider{})
+		require.NoError(t, err)
+		defer stop(t, svc)
+
+		// register action event channel
+		aCh := make(chan service.DIDCommAction)
+		require.NoError(t, svc.RegisterActionEvent(aCh))
+
+		// register action event channel
+		sCh := make(chan service.StateMsg)
+		require.NoError(t, svc.RegisterMsgEvent(sCh))
+
+		// creates threadID
+		thID := uuid.New().String()
+
+		// creates proposal msg
+		reqMsg, err := service.NewDIDCommMsg(toBytes(t, Proposal{
+			Type: ProposalMsgType,
+			ID:   thID,
+		}))
+		require.NoError(t, err)
+
+		// handle Proposal msg (sends Request)
+		go func() {
+			// nolint: govet
+			_, err := svc.HandleInbound(reqMsg)
+			require.NoError(t, err)
+		}()
+
+		dep := mocks.NewMockActionDependency(ctrl)
+		dep.EXPECT().Destinations().Return(nil)
+
+		continueAction(t, aCh, ProposalMsgType, dep)
+		checkStateMsg(t, sCh, service.PreState, ProposalMsgType, stateNameDeciding)
+		checkStateMsg(t, sCh, service.PostState, ProposalMsgType, stateNameDeciding)
+		checkStateMsg(t, sCh, service.PreState, ProposalMsgType, stateNameWaiting)
+		checkStateMsg(t, sCh, service.PostState, ProposalMsgType, stateNameWaiting)
+
+		// creates Ack msg
+		ackMsg, err := service.NewDIDCommMsg(toBytes(t, model.Ack{
+			Type: AckMsgType,
+			ID:   thID,
+		}))
+		require.NoError(t, err)
```

Comment:
this is a general question, not just specific to this line:

I don't see error (failure test cases in the tests), there's only require.NoError but no require.Error... 

Adding them will increase the test coverage as well. As a general rule we try to increase the code coverage in the project with each patch.

Reply:
in this package, only one line is not covered
```
data, err := json.Marshal(o)
if err != nil {
	return fmt.Errorf("marshal options: %w", err)
}
```

Reply:
Unless `o` can be injected.. then it's fine as we have no control over JSON marshalling a struct.